### PR TITLE
Use "current state" when `null` is passed to `go`/`makePath`

### LINKED ISSUE #83: State name should be optional to `state.go`
If the first argument to `state.go` is not a string, then it should behave as if you had passed in the name of the current state.

**Discussion point**: should the "current state" be the last state that the router started transitioning to, or should it be the last state that was fully loaded?

Put another way, if you call `redirect({ someParameter: 'some other option' })` inside a `resolve` function, should it redirect you to the state that was in the middle of being loaded, or the state before that?

I'm leaning towards the state that you're in the middle of transitioning to, but I'm open to other arguments.

### PR DESCRIPTION
**Pull request by TehShrike**
Fixes #83 

Want to review @daytonlowell and @ArtskydJ ?

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -308,8 +308,10 @@ module.exports = function StateProvider(makeRenderer, rootElement, stateRouterOp
 			parameters = extend(lastStateStartedActivating.get().parameters, parameters)
 		}
 
-		prototypalStateHolder.guaranteeAllStatesExist(stateName)
-		var route = prototypalStateHolder.buildFullStateRoute(stateName)
+		var destinationState = stateName === null ? lastStateStartedActivating.get().name : stateName
```
Comment (ArtskydJ) [inline index.js]: We're sure that `lastStateStartedActivating.get()` will always return a state? What happens if in your code, you do a `state.go(null)` when you had no "previous state"?

Comment (TehShrike) [inline index.js]: Good question.  We should probably have a test for that, and throw an error with a useful message?

**Comment by ArtskydJ** [comment]
LGTM, aside from a few test variable names being a bit juvenile. :confused: But it's not my code base, and I guess there were already some silly names, so if you're ok with it, go ahead and merge! :grinning:

**Comment by daytonlowell** [comment]
Looks good to me. Sorry, I totally forgot about makePath and didn't make a test for it.

**Comment by TehShrike** [comment]
No problem, I forgot about it too until I went to go write the implementation itself.

**Comment by TehShrike** [comment]
I pushed up a test and error for the case where there was no previous state to look at.

Before the change, the error thrown was `State  does not exist`.

Thoughts?

File: index.js [hunk 2]
```diff
@@ -304,11 +304,17 @@ module.exports = function StateProvider(makeRenderer, rootElement, stateRouterOp
...
+				throw new Error('makePath required a previous state to exist, and none was found')
+			}
+		}
 		if (options && options.inherit) {
+			guaranteePreviousState()
 			parameters = extend(lastStateStartedActivating.get().parameters, parameters)
 		}
 
-		var destinationState = stateName === null ? lastStateStartedActivating.get().name : stateName
+		var destinationState = stateName === null ? (guaranteePreviousState(), lastStateStartedActivating.get().name) : stateName
```
Comment (ArtskydJ) [inline index.js]: IMO, since the comma operator is so obscure, it makes the code harder to read.

File: index.js [hunk 3]
```diff
@@ -304,11 +304,17 @@ module.exports = function StateProvider(makeRenderer, rootElement, stateRouterOp
 	}
 
 	function makePath(stateName, parameters, options) {
+		function guaranteePreviousState() {
+			if (!lastStateStartedActivating.get().name) {
+				throw new Error('makePath required a previous state to exist, and none was found')
+			}
+		}
```
Comment (ArtskydJ) [inline index.js]: Getting into semantics here, but does `guaranteePreviousState()` actually gaurantee that a previous states exists? I feel like it's more of an assertion.

**Comment by ArtskydJ** [comment]
Looks like I commented on an outdated diff. Your newest commit(s) fixed the issues I brought up.

If that error can be thrown from a `router.go()` call, then the new message might confuse people, since "makePath" is referenced.

Other than that, LGTM!

**Comment by TehShrike** [comment]
Published as 5.13.0

**Pull request merged**